port/raspberrypi/rp2xxx/src/hal/dma.zig

```diff
@@ -102,13 +130,201 @@ pub const Channel = enum(u4) {
+                }
+            }
+
+            inline fn get_count(value: anytype) u32 {
+                const Type = @TypeOf(value);
+                const Info = @typeInfo(Type);
+                switch (Info) {
+                    .pointer => |ptr| {
+                        switch (ptr.size) {
+                            .one => switch (@typeInfo(ptr.child)) {
+                                .array => |array| {
+                                    return array.len;
+                                },
+                                else => return 1,
+                            },
+                            .many, .slice => return value.len,
+                            .c => unreachable,
+                        }
+                    },
+                    else => unreachable,
+                }
+            }
+
+            inline fn get_increment(Type: type) bool {
+                const Info = @typeInfo(Type);
+                return switch (Info) {
+                    .pointer => |ptr| switch (ptr.size) {
+                        .one => switch (@typeInfo(ptr.child)) {
+                            .array => true,
+                            else => false,
+                        },
+                        .many, .slice => true,
+                        .c => unreachable,
+                    },
+                    else => comptime if (is_peripheral(Type)) false else unreachable,
+                };
+            }
+
+            fn type_to_data_size(Type: type) ?DataSize {
+                return switch (Type) {
+                    u8, i8 => .size_8,
+                    u16, i16 => .size_16,
+                    u32, i32 => .size_32,
+                    else => null,
+                };
+            }
+
+            fn get_data_size(Type: type) ?DataSize {
+                // at this point we are guarnteed that we have pointer type
+                const Info = @typeInfo(Type);
+                const ChildType = Info.pointer.child;
+                return switch (@typeInfo(ChildType)) {
+                    .array => |array| type_to_data_size(array.child),
+                    .int => type_to_data_size(ChildType),
+                    else => null,
+                };
+            }
+        };
+
+        const WriteType = @TypeOf(write);
+        const ReadType = @TypeOf(read);
+
+        comptime H.validate_type(WriteType);
+        comptime H.validate_type(ReadType);
+
+        const write_addr = H.get_addr(write);
+        const read_addr = H.get_addr(read);
+
+        comptime if (H.is_peripheral(ReadType) and H.is_peripheral(WriteType))
+            @compileError("cross peripheral dma is unsupported");
+
+        const data_size = comptime if (H.is_peripheral(WriteType))
+            H.get_data_size(ReadType).?
+        else
+            H.get_data_size(WriteType).?;
+
+        const dreq = if (comptime H.is_peripheral(WriteType)) H.get_dreq(write) else H.get_dreq(read);
+
+        const count = blk: {
```

Comment:
could make this whole block comptime to make it clear

Reply:
i can't because for slices it depends on runtime length